packages/firestore/src/remote/online_state_tracker.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OnlineState } from '../core/types';
+import * as log from '../util/log';
+import { assert } from '../util/assert';
+
+const LOG_TAG = 'OnlineStateTracker';
+
+// To deal with transient failures, we allow multiple stream attempts before
+// giving up and transitioning to Offline.
+const MAX_WATCH_STREAM_FAILURES = 2;
+
+// To deal with stream attempts that don't succeed or fail in a timely manner,
+// we have a maximum timeout we'll wait for the stream to either succeed or fail
+// MAX_WATCH_STREAM_FAILURES times, else we revert to OnlineState.Offline.
+const MAX_WATCH_STREAM_TIMEOUT_MS = 10 * 1000;
+
+/**
+ * A component used by the RemoteStore to track the OnlineState (that is,
+ * whether or not the client as a whole should be considered to be online or
+ * offline), implementing the appropriate heuristics.
+ *
+ * In particular, when the client is trying to connect to the backend, we
+ * allow up to MAX_WATCH_STREAM_FAILURES within MAX_WATCH_STREAM_TIMEOUT_MS for
+ * a connection to succeed. If we have too many failures or the timeout elapses,
+ * then we set the OnlineState to Offline, and the client will behave as if
+ * it is offline (get()s will return cached data, etc.).
+ */
+export class OnlineStateTracker {
```

Comment:
I can't think of any way it would happen but just in case: are there any interactions between this class and the idle timeout timers?

Reply:
There shouldn't be since this timer goes away once you're connected. And the idle timers start once you're connected...  And with the improved testing, if both timers were active at once, runDelayedOperationsEarly() would run both so we'd be exercising them together, at least in theory...